3drpp_appx/src/fmm.cpp

```diff
@@ -317,25 +317,76 @@ void rtfmm::LaplaceFMM::init_reg_body(Cells3& cells)
         printf("search 0,0,0\n");
         // search reg body in 0,0,0
         printf("~~~~~~~~~~ rega!\n");
-        for(int i = 0; i < bs.size(); i++)
+
+        std::vector<Indices> boundary_bodies_idxs(leaves.size());
+        // find weights for existing bodies within the same cell
+        for(int leaf_idx = 0; leaf_idx < leaves.size(); leaf_idx++)
         {
-            vec3r ploc = bs[i].x;
-            for(int leaf_idx = 0; leaf_idx < leaves.size(); leaf_idx++)
+            Cell3& cell = cells[leaves[leaf_idx]];
+            cell.weights.resize(cell.brange.number);
+            cell.bodies.resize(cell.brange.number);
+            int rx = 0;
+            for(int body_idx = cell.brange.offset; body_idx < cell.brange.offset + cell.brange.number; body_idx++)
             {
-                Cell3& cell = cells[leaves[leaf_idx]];
-                if(!(i >= cell.brange.offset && i < cell.brange.offset + cell.brange.number))
+                const Body3 &body = bs[body_idx];
+                const vec3r dx = body.x - cell.x;
+                const vec3r dx_simcenter =
+                    (body.x - args.x).abs() + args.rega;
+                vec3r ws = get_w_xyz(dx, cell.r, args.rega);
+
+                for (int d = 0; d <= 2; d++)
                 {
-                    vec3r dx = ploc - cell.x;
-                    real w = get_w_xyz(dx, cell.r, args.rega).mul();
-                    if(w > 0)
+                    if (dx_simcenter[d] >= args.r)
                     {
-                        cell.reg_body_idx.push_back(std::make_pair(i, vec3r(0,0,0)));
-                        //printf("push %d -> %d\n", i, leaf_idx);
+                        ws[d] = 1;
                     }
                 }
+                const real w = ws.mul();
+
+                if (w < 1)
+                {
+                    boundary_bodies_idxs[leaf_idx].push_back(body_idx);
+                }
+
+                cell.bodies[rx] = body;
+                cell.weights[rx++] = w;
```

Comment:
In the case image > 0, if a body `i` is located at the corner of a cell `j`, according to your code, its `w` will be 1, which means it contributes to `j` completely.
But according to the definition of regularization, it should be shared by 8 cells.

The reason why i insert `if(args.images == 0)` is to ensure `i` contributes to `j` completely when PBC is not used, but shared by cells when PBC is used.

So I think `if(args.images == 0)` should be added here.

Reply:
I'm curious about the l2p error.

Reply:
I see, I updated it as per your suggestion.

Please see if there is anything else to be modified.

Reply:
I see, but I think another aspect is with uniform trees if particle distribution is sparse or dense in certain regions then we end up with a lot of empty cells or very huge cells compared to others in the tree. 

I tried an [example](http://www.mdtutorials.com/gmx/lysozyme/index.html) in GROMACS which has such a distribution. 

So for the time being, we keep this a choice. 

As for current change, it is independent of the choice of tree, it can work with both.

Reply:
I see
Let's key it a choice.
I think now we can merge.